harness/determined/core/_train.py

```diff
@@ -48,6 +68,9 @@ def set_status(self, status: str) -> None:
         logger.debug(f"set_status({status})")
         self._session.post(f"/api/v1/trials/{self._trial_id}/runner/metadata", json=body)
 
+    def set_tensorboard_metric_writer(self, writer: tensorboard.BatchMetricWriter) -> None:
+        self._tbd_writer = writer
+
```

Comment:
This is a form of lazy initialization, and it makes it harder to think about the object because of the ways it can mutate over time.

Let's get rid of it and keep the API behavior simpler.  I think either you set up Auto mode and we picked a metric writer for you, or you set up Manual mode and we shouldn't allow you to set the metric writer afterwards.

Reply:
It sure _is_ lazy initialization. The reason I have to use it is that we have not instantiated a trial/controller pair at the moment of `core.init()` and therefore do not know which metric writer we intend to use.
The order of initialization is:
`core.init() -> core_context -> trial_context -> trial -> controller -> metric_writer -> core_context.set_tensorboard_metric_writer`. The alternative that avoids lazy initialization is what I call the "original kludge": to step on a potential land mine, and if we explode then use a second life to instantiate another writer (with no guarantee that we will actually be using the correct writer). Is there another alternative?